flight/Libraries/math/statistics.c

```diff
@@ -0,0 +1,352 @@
+ * the vehicle is still flying.
+ * @param p_sense_condition_H0 The probability of sensing the H0 condition, given the real condition
+ * @param p_sense_condition_H1 The probability of sensing the H1 condition, given the real condition
+ * @param p_H0_Ux_H0 The probability of state H0 staying H0 after the control is applied
+ * @param p_H0_Ux_H1 The probability of state H1 becoming H0 after the control is applied
+ * @param p_H1_Ux_H0 The probability of state H0 becoming H1 after the control is applied
+ * @param p_H1_Ux_H1 The probability of state H1 staying H1 after the control is applied
+ */
+void bayes_filter(double *belief_H0, double *belief_H1,
+                double p_sense_condition_H0, double p_sense_condition_H1,
+                double p_H0_Ux_H0, double p_H0_Ux_H1,
+                double p_H1_Ux_H0, double p_H1_Ux_H1)
+{
+	// --------------------//
+	// Process the control //
+	// ------------------- //
+	// Helper variable to let the function read better
+	double old_belief_H0 = *belief_H0;
+	double old_belief_H1 = *belief_H1;
+
+	// The median belief is the old belief modified by the likelihood that the
+	// control successfully controls the measured state
+	double belief_bar_H0 = p_H0_Ux_H0 * old_belief_H0 + p_H0_Ux_H1 * old_belief_H1;
+	double belief_bar_H1 = p_H1_Ux_H0 * old_belief_H0 + p_H1_Ux_H1 * old_belief_H1;
+
+	// -------------------//
+	// Measurement update //
+	// -------------------//
+	// Helper variable to let the function read better
+	double *new_belief_H0 = belief_H0;
+	double *new_belief_H1 = belief_H1;
+
+	// The new belief is the likelihood of the new state occurring times the
+	// likelihood of measuring the new state
+	*new_belief_H0 = p_sense_condition_H0 * belief_bar_H0;
+	*new_belief_H1 = p_sense_condition_H1 * belief_bar_H1;
+
+
+	// Renormalize by gamma. This keeps the sum of the two values at 1, i.e. 100%
+	// of probable cases covered
+	double gamma = 1/(*new_belief_H0 + *new_belief_H1);
+	*new_belief_H0 *= gamma;
+	*new_belief_H1 *= gamma;
+
+}
+
+
+/**
+ * @brief initialize_linear_sums Because floating point values are used,
+ * there will be some loss of precision. Thus, it is important to periodically reset the parameters.
+ * Cf. https://en.wikipedia.org/wiki/Algorithms_for_calculating_variance
+ * Cf. http://math.stackexchange.com/questions/102978/incremental-computation-of-standard-deviation
+ * @param X struct holding the statistical parameters
+ * @param window_size desired window size
+ * @param num_samples number of samples available in vector
+ * @param y vector of samples
+ */
+void initialize_linear_sums(struct linear_mean_and_std_dev *X, uint16_t window_size, uint16_t num_samples, const float y[])
+{
+	X->window_size = window_size;
+	X->T0 = num_samples;
+
+	// Reinitialize the accumulators
+	X->T1 = 0;
+	X->T2 = 0;
+
+	for (int i=0; i<num_samples; i++) {
+		// Sum of elements
+		X->T1 += (double)y[i];
+
+		// Sum of squares of elements
+		X->T2 += (double)y[i]*(double)y[i];
```

Comment:
Yes, 1.0 * ... will promote to double though without a cast (and without the downsides of casting).  OTOH it's slightly more arcane.